File: java/tst/tarehart/rlbot/tuning/AccelerationPredictedVsActualTest.java
package tarehart.rlbot.tuning;

import org.junit.Assert;
import org.junit.Test;
import tarehart.rlbot.carpredict.CarPath;
import tarehart.rlbot.carpredict.CarSlice;
import tarehart.rlbot.math.DistanceTimeSpeed;
import tarehart.rlbot.physics.DistancePlot;
import tarehart.rlbot.time.Duration;
import tarehart.rlbot.time.GameTime;

import java.io.UnsupportedEncodingException;
import java.util.List;
import java.util.Optional;

public class AccelerationPredictedVsActualTest {

    private CarPathHelper carPathHelper = new CarPathHelper();

    @Test
    public void fullBoost() throws UnsupportedEncodingException {
        testFile("fullBoostAccel.json", true);
    }

    @Test
    public void zeroBoost() throws UnsupportedEncodingException {
        testFile("zeroBoostAccel.json", false);
    }


    private void testFile(String filename, boolean hasBoost) throws UnsupportedEncodingException {

        CarPath actualPath = carPathHelper.readRecording(filename);
        DistancePlot predictedAccel = carPathHelper.makePrediction(actualPath, hasBoost);
        // (-73.29997, 65.447556, 4.5342107) after first time step

        List<CarSlice> actual = actualPath.getSlices();
        GameTime startTime = actual.get(0).time;

        for (CarSlice slice : actual) {
            Optional<DistanceTimeSpeed> predicted = predictedAccel.getMotionAfterDuration(Duration.Companion.between(startTime, slice.time));
            Assert.assertTrue(predicted.isPresent());
            Assert.assertEquals(predicted.get().getSpeed(), slice.getVelocity().magnitude(), 1);
        }
    }





}

File: src/test/java/tarehart/rlbot/tuning/AccelerationPredictedVsActualTest.kt
package tarehart.rlbot.tuning

import org.junit.Assert
import org.junit.Test
import tarehart.rlbot.carpredict.CarPath
import tarehart.rlbot.carpredict.CarSlice
import tarehart.rlbot.math.DistanceTimeSpeed
import tarehart.rlbot.physics.DistancePlot
import tarehart.rlbot.time.Duration
import tarehart.rlbot.time.GameTime

import java.io.UnsupportedEncodingException
import java.util.Optional

class AccelerationPredictedVsActualTest {

    private val carPathHelper = CarPathHelper()

    @Test
    @Throws(UnsupportedEncodingException::class)
    fun fullBoost() {
        testFile("fullBoostAccel.json", true)
    }

    @Test
    @Throws(UnsupportedEncodingException::class)
    fun zeroBoost() {
        testFile("zeroBoostAccel.json", false)
    }


    @Throws(UnsupportedEncodingException::class)
    private fun testFile(filename: String, hasBoost: Boolean) {

        val actualPath = carPathHelper.readRecording(filename)
        val predictedAccel = carPathHelper.makePrediction(actualPath, hasBoost)
        // (-73.29997, 65.447556, 4.5342107) after first time step

        val actual = actualPath.path
        val startTime = actual[0].time

        for (slice in actual) {
            val predicted = predictedAccel.getMotionAfterDuration(Duration.between(startTime, slice.time))
            predicted ?: throw AssertionError()
            Assert.assertEquals(predicted.speed, slice.velocity.magnitude(), 1F)
        }
    }
}
